core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServices.java

```diff
@@ -310,42 +314,126 @@ private static String findPattern(Method method, Set<Annotation> methodAnnotatio
     }
 
     /**
-     * Returns a decorator chain which is specified by {@link Decorator} annotations.
+     * Returns a decorator chain which is specified by {@link Decorator} annotations and user-defined
+     * decorator annotations.
      */
     private static Function<Service<HttpRequest, HttpResponse>,
             ? extends Service<HttpRequest, HttpResponse>> decorator(Method method, Class<?> clazz) {
 
-        // Class-level decorators are applied before method-level decorators.
-        final Function<Service<HttpRequest, HttpResponse>,
-                ? extends Service<HttpRequest, HttpResponse>> decorator =
-                decorator(clazz.getAnnotationsByType(Decorator.class),
-                          decorator(method.getAnnotationsByType(Decorator.class), null));
+        final List<DecoratorAndOrder> decorators = collectDecorators(clazz, method);
 
+        Function<Service<HttpRequest, HttpResponse>,
+                ? extends Service<HttpRequest, HttpResponse>> decorator = null;
+        Collections.reverse(decorators);
+        for (final DecoratorAndOrder d : decorators) {
```

Comment:
I don't know if it is a good idea to go the loop from the back so that we don't have to do the `reverse()` here.

```
for (int i = decorators.size() - 1; i >= 0; i--) {
    decorator = decorator == null ? decorators.get(i).decorator()
                                  : decorator.andThen(decorators.get(i).decorator());
}
```
But it looks like more verbose than the codes above. So If you don't like it, just ignore this.:-)

Reply:
I also considered that kind of for-loop. Let me change it.